### ISSUE TITLE
Datadog integration: NaN for gaugeCollectionSize

### ISSUE THREAD

**Issue by mfernandezpajares** [micrometer-metrics/micrometer#400]
Hi all!

we are using this metric:

```
Metrics.gaugeCollectionSize(
                "some.metric",
                listOf(Tag.of("sometag", "tag.value")),
                someNonNullList)
```

and someNonNullList can never be null, and the situation is that in Datadog the metric disappears but spring-boot metrics endpoint reports `NaN` as value. Though this behaviour is not consistent, sometimes the values is there and sometimes is not.

This problem was not happening when we were using Micrometer with Statsd

**Comment by jkschneider** [comment]
@mfernandezpajares Cool, first Github issue!

Datadog's implementation uses `DefaultGauge`, whose implementation only reports NaN if the monitored object becomes null, as you can see [in this condition](https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/internal/DefaultGauge.java#L40).

`DefaultGauge` also intentionally holds a weak reference on the object being monitored, so the list could get garbage collected if it is otherwise dereferenced. When it gets garbage collected, trying to retrieve the value from the weak reference returns `null`, and the gauge starts reporting NaN. Can you check that you are holding a strong reference to the list?

I was able to reproduce the behavior you describe in this Gist if I manually dereference the list. At some point a garbage collection event happens against the eden space and the Gauge flips to 
NaN. At that point, I can observe how Datadog's plot drops off, as expected:

![image](https://user-images.githubusercontent.com/1697736/35878187-1d54c2ee-0b3d-11e8-9319-ee3981a2ba54.png)

**Comment by jkschneider** [comment]
Fixed in #406, I think.

**Comment by mfernandezpajares** [comment]
Hi!! sorry for the late response and thanks a lot for your response, it was very helpful! 
Indeed we use weak references. We will try then traditional way or reporting collection size through normal gauge.
